contentcuration/contentcuration/utils/publish.py

```diff
@@ -819,62 +817,74 @@ def fill_published_fields(channel, version_notes):
     channel.save()
 
 
-def create_or_update_tsvectors(channel_id):
+def sync_contentnode_and_channel_tsvectors(channel_id):
     """
-    Create or update tsvectors for the channel and all its content nodes.
+    Creates, deletes and updates tsvectors of the channel and all its content nodes
+    to reflect the current state of channel's main tree.
     """
     # Update or create channel tsvector entry.
     logging.info("Starting to set tsvectors for channel with id {}.".format(channel_id))
 
     from contentcuration.viewsets.channel import primary_token_subquery
 
     channel = (ccmodels.Channel.objects
+               .filter(pk=channel_id)
                .annotate(primary_channel_token=primary_token_subquery,
                          keywords_tsvector=CHANNEL_KEYWORDS_TSVECTOR)
-               .get(pk=channel_id))
+               .values("keywords_tsvector", "main_tree__tree_id")
+               .get())
 
     if ChannelFullTextSearch.objects.filter(channel_id=channel_id).exists():
-        update_count = ChannelFullTextSearch.objects.filter(channel_id=channel_id).update(keywords_tsvector=channel.keywords_tsvector)
+        update_count = ChannelFullTextSearch.objects.filter(channel_id=channel_id).update(keywords_tsvector=channel["keywords_tsvector"])
         logging.info("Updated {} channel tsvector.".format(update_count))
     else:
-        obj = ChannelFullTextSearch(channel_id=channel_id, keywords_tsvector=channel.keywords_tsvector)
+        obj = ChannelFullTextSearch(channel_id=channel_id, keywords_tsvector=channel["keywords_tsvector"])
         obj.save()
         logging.info("Created 1 channel tsvector.")
 
     # Update or create contentnodes tsvector entry for channel_id.
-    logging.info("Setting tsvectors for all contentnodes in channel {}.".format(channel_id))
+    logging.info("Starting to set tsvectors for all contentnodes in channel {}.".format(channel_id))
 
-    if ContentNodeFullTextSearch.objects.filter(channel_id=channel_id).exists():
+    nodes_tsvector_query = (ccmodels.ContentNode.objects
+                            .filter(tree_id=channel["main_tree__tree_id"])
+                            .annotate(channel_id=Value(channel_id),
+                                      contentnode_tags=StringAgg("tags__tag_name", delimiter=" "),
+                                      keywords_tsvector=CONTENTNODE_KEYWORDS_TSVECTOR,
+                                      author_tsvector=CONTENTNODE_AUTHOR_TSVECTOR)
+                            .order_by())
 
```

Comment:
I think the most memory efficient way to do this might be to first do a bulk create just to create the table entries for each content node:

```
ContentNodeFullTextSearch.objects.bulk_create([ContentNodeFullTextSearch(contentnode_id=c_id, channel_id=channel_id) for c_id in nodes_ts_vector_query.values_list("id", flat=True)])
```

You can then do a correlated subquery for the update:
```
cn_subquery = ccmodels.ContentNode.objects.filter(id=OuterRef("contentnode_id"))

ContentNodeFullTextSearch.objects.filter(channel_id=channel_id).update(
  keywords_tsvector=Subquery(cn_subquery.annotate(
    contentnode_tags=StringAgg("tags__tag_name", delimiter=" "),
    keywords_tsvector=CONTENTNODE_KEYWORDS_TSVECTOR
  ).order_by().values_list("keywords_tsvector")[:1]),
  author_tsvector=Subquery(cn_subquery.annotate(
    author_tsvector=CONTENTNODE_AUTHOR_TSVECTOR
  ).order_by().values_list("author_tsvector")[:1])
)
```

This should have the advantage of all the correlated updates for the TSVectors being left to the database and not being brought into Python memory. Only the list of affected ContentNode ids is ever read out of the database.

Reply:
I am assuming that you expect to first delete all nodes with `channel_id` like `ContentNodeFullTextSearch.objects.filter(channel_id=channel_id).delete()` and then you are writing the above queries. 

```
ContentNodeFullTextSearch.objects.bulk_create([ContentNodeFullTextSearch(contentnode_id=c_id, channel_id=channel_id) for c_id in nodes_ts_vector_query.values_list("id", flat=True)])
```

The above is creating a list of all the `ContentNodeFullTextSearch` objects which will consume a lot of memory. I think sir the memory will be consumed until the list is fully evaluated and then it'll be immediately garbage collected I guess...? So we are indeed consuming the same amount of memory as the current implementation but with the advantage of decreased consumption time. Is my understanding correct sir?   

Since, we are deleteing all and creating all the objects, we can simply add tsvectors while insertion itself. It'll be the most optimal way in terms of memory use and performance ONLY IF we are first deleting all the nodes. Here's my concern regarding deletion of all nodes quoted from my previous comment: 

> In delete all, insert all there's a performance downside. For e.g. if the user has 10,000 nodes and he has added 10 more nodes & updated 100 nodes and then re-publishes, it'll delete & re-insert the already up to date 9,900 tsvectors. Keeping in mind the updates of GIN indexes and heavy cost of delete and insert, I decided to move ahead with the current implementation.



```
cn_subquery = ccmodels.ContentNode.objects.filter(id=OuterRef("contentnode_id"))

ContentNodeFullTextSearch.objects.filter(channel_id=channel_id).update(
  keywords_tsvector=Subquery(cn_subquery.annotate(
    contentnode_tags=StringAgg("tags__tag_name", delimiter=" "),
    keywords_tsvector=CONTENTNODE_KEYWORDS_TSVECTOR
  ).order_by().values_list("keywords_tsvector")[:1]),
  author_tsvector=Subquery(cn_subquery.annotate(
    author_tsvector=CONTENTNODE_AUTHOR_TSVECTOR
  ).order_by().values_list("author_tsvector")[:1])
)
```

I tried updating the nodes with a very similar update query like you've written but the problem is that it sets all the nodes with the same value or NULL. This is because the `UPDATE SET WHERE` SQL is meant to set all of the records with the same value. 

I tried to fix the above problem by joining the `ContentNodeFullTextSearch` and `ContentNode` table and then updating the records to `CONTENTNODE_KEYWORDS_TSVECTOR` but Django doesn't allows updates on joined table due to internal safety reasons which I'm unaware of.

<hr>

@rtibbles sir in my opinion we should go ahead with updates and inserts in chunks. That would be a win from all sides. Memory use will be under check and performance will be optimal as well.

What do you think sir?

Reply:
I like the correlated subquery approach @rtibbles mentioned and I think it goes along with we talked about on slack, Vivek, about minimizing the queries needed to create/update the models to just a handful, since everything needed to make these updates is contained within the database already. Backing out into python for processing these should be unnecessary.

Reply:
I think we can probably avoid the overhead of deleting everything to start with, yes, so maybe we do just do the selective deletion of `ContentNodeFullTextSearch` for no longer relevant nodes, and creation of new `ContentNodeFullTextSearch` for newly published nodes, then do the update subquery?

Reply:
Yes, so let us move forward with selective delete, updates to existing nodes and then insert in chunks.